src/NHibernate/Persister/Collection/AbstractCollectionPersister.cs

```diff
@@ -1501,10 +1501,41 @@ protected virtual string FilterFragment(string alias)
 			return HasWhere ? " and " + GetSQLWhereString(alias) : "";
 		}
 
+		/// <summary>
+		/// Get the where clause filter, given a query alias and considering enabled session filters.
+		///
+		/// When an <see cref="IEntityPersister"/> is set, the filter generated by the persister is returned. This ensures
+		/// that a proper filter is also generated for subclass hierarchy relations.
+		/// </summary>
+		/// <param name="alias"></param>
+		/// <param name="enabledFilters"></param>
+		/// <returns></returns>
 		public virtual string FilterFragment(string alias, IDictionary<string, IFilter> enabledFilters)
 		{
+			if(!enabledFilters.Any())
+			{
+				// no filters enabled
+				// but a potential where condition needs to be handled properly
+				return FilterFragment(alias);
+			}
+			
+			if(!FilterHelper.IsAffectedBy(enabledFilters))
+			{
+				// the collection filters don't match with the enabled filters
+				// but a potential where condition needs to be handled properly
+				return FilterFragment(alias);
+			}
+			
+			if(ElementPersister != null && 
+			   ElementPersister.FilterHelper.IsAffectedBy(enabledFilters))
```

Comment:
`ElementPersister.FilterHelper.IsAffectedBy(enabledFilters)` check looks wrong and should be removed. I supect that's the reason for failing test with separate collection filter

Reply:
I've decoupled the generation of the filter code from the EntityPersister now. The EntityPersister just provides the column to table alias mapping anymore. So for collections and entities, the same logic to resolve the proper table alias for the filter generation is used.